[GPU Process] Possible null dereferencing when destroying the cached RenderingResources of a WebPage

shallawa opened the pull request:
#### a71e93715498ea8ea6854bb921d20646b5f99ccc
<pre>
[GPU Process] Possible null dereferencing when destroying the cached RenderingResources of a WebPage
<a href="https://bugs.webkit.org/show_bug.cgi?id=267388">https://bugs.webkit.org/show_bug.cgi?id=267388</a>
<a href="https://rdar.apple.com/120464355">rdar://120464355</a>

Reviewed by Chris Dumez.

Because RemoteResourceCacheProxy stores the cached RenderingResources in a HashMap
of WeakPtrs, RemoteResourceCacheProxy::clearRenderingResourceMap() needs to ensure
the WeakPtr of the RenderingResource is not null before using it.

* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearRenderingResourceMap):

Canonical link: <a href="https://commits.webkit.org/272919@main">https://commits.webkit.org/272919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c2ff107168ca825a109b983eff5fc6194e94d907

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12309 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36156 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30438 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14665 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9468 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29551 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34013 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10361 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/29986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9056 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9158 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37484 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30428 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/30316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35289 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7245 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33165 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11060 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7769 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9880 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10050 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -229,8 +229,10 @@ void RemoteResourceCacheProxy::releaseRenderingResource(RenderingResourceIdentif
 {
-    for (auto& renderingResource : m_renderingResources.values())
-        renderingResource.get()->removeObserver(*this);
+    for (auto& renderingResource : m_renderingResources.values()) {
```
I would rename `renderingResource` to `weakRenderingResource` here.

cdumez commented on Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
```diff
@@ -229,8 +229,10 @@ void RemoteResourceCacheProxy::releaseRenderingResource(RenderingResourceIdentif
-    for (auto& renderingResource : m_renderingResources.values())
-        renderingResource.get()->removeObserver(*this);
+    for (auto& renderingResource : m_renderingResources.values()) {
+        if (auto renderingResourcePtr = renderingResource.get())
```
Then use `RefPtr renderingResource = weakRenderingResource.get()` here.

Besides the naming, note that we now prefer to use smart pointers instead of `auto` for clarity.

cdumez reviewed (approved):
LGTM with nits.

shallawa committed:
[GPU Process] Possible null dereferencing when destroying the cached RenderingResources of a WebPage
https://bugs.webkit.org/show_bug.cgi?id=267388
rdar://120464355

Reviewed by Chris Dumez.

Because RemoteResourceCacheProxy stores the cached RenderingResources in a HashMap
of WeakPtrs, RemoteResourceCacheProxy::clearRenderingResourceMap() needs to ensure
the WeakPtr of the RenderingResource is not null before using it.

* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::clearRenderingResourceMap):

Canonical link: https://commits.webkit.org/272919@main
